File: lol/src/test/kotlin/com/your/lol/domain/LoLScoreStatsCalculatorTest.java
package com.your.lol.domain;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.offset;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;

import com.google.gson.reflect.TypeToken;
import com.your.lol.dto.riot.MatchDto;
import com.your.lol.dto.statistics.ChampionDataDto;
import com.your.lol.dto.statistics.StatsDto;
import com.your.lol.support.JsonFileConverter;
import com.your.lol.support.WebClientFacade;

import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;

class LoLScoreStatsCalculatorTest {

    private WebClientFacade webClientFacade = mock(WebClientFacade.class);
    private LoLScoreStatsCalculator loLScoreStatsCalculator = new LoLScoreStatsCalculator(webClientFacade);


    @Test
    void calculateStatisticsBySummonerName() {
        // given
        List<MatchDto> matchDtos = createTestData();
        given(webClientFacade.requestRiotStatistics("테스트 소환사 GG"))
                .willReturn(matchDtos);

        // when
        StatsDto result = loLScoreStatsCalculator.calculateStatsBySummonerName("테스트 소환사 GG");
        ChampionDataDto Jinx = findChampionByName(result, "Jinx");
        ChampionDataDto Riven = findChampionByName(result, "Riven");

        // then
        assertAll(
                () -> assertThat(result.getWinRate()).isCloseTo(0.66, offset(0.01)),
                () -> assertThat(Jinx.getMatchCount()).isEqualTo(1),
                () -> assertThat(Jinx.getWinRate()).isEqualTo(1.0),
                () -> assertThat(Jinx.getAverageKda()).isEqualTo(1.0),
                () -> assertThat(Riven.getMatchCount()).isEqualTo(2),
                () -> assertThat(Riven.getWinRate()).isEqualTo(0.5),
                () -> assertThat(Riven.getAverageKda()).isEqualTo(1.0)
        );
    }

    @NotNull
    private static ChampionDataDto findChampionByName(StatsDto stats, String championName) {
        return stats.getChampionDataDtos().stream()
                .filter(it -> it.getChampionName().equals(championName))
                .findAny()
                .get();
    }

    private List<MatchDto> createTestData() {
        return JsonFileConverter.fromJsonFile(
                "match-dtos.json",
                new TypeToken<>() {
                });
    }
}

File: lol/src/test/kotlin/com/your/lol/domain/LoLScoreStatsCalculatorTest.kt
package com.your.lol.domain

import com.your.lol.dto.riot.MatchDto
import com.your.lol.dto.statistics.ChampionDataDto
import com.your.lol.dto.statistics.StatsDto
import com.your.lol.support.GlobalTestUtils
import com.your.lol.support.WebClientFacade
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class LoLScoreStatsCalculatorTest : GlobalTestUtils() {

    private val webClientFacade: WebClientFacade = mockk()
    private val loLScoreStatsCalculator: LoLScoreStatsCalculator = LoLScoreStatsCalculator(webClientFacade)


    @Test
    fun `소환사 이름을 검색하여 전적 통계를을 산출한다`() {
        // given
        val matchDtos: List<MatchDto> = createTestData("match-dtos.json")
        every { webClientFacade.requestRiotStatistics("테스트 소환사 GG") } returns matchDtos

        // when
        val statsDto: StatsDto = loLScoreStatsCalculator.calculateStatsBySummonerName("테스트 소환사 GG")
        val jinx: ChampionDataDto = findChampionByName(statsDto, "Jinx")
        val riven: ChampionDataDto = findChampionByName(statsDto, "Riven")

        // then
        Assertions.assertAll(
                assertCloseTo(statsDto.winRate, 0.666, 0.001),
                assertEquals(jinx.matchCount, 1),
                assertEquals(jinx.winRate, 1.0),
                assertCloseTo(jinx.averageKda, 3.6, 0.01),
                assertEquals(riven.matchCount, 2),
                assertEquals(riven.winRate, 0.5),
                assertCloseTo(riven.averageKda, 5.97, 0.01)
        )
    }
}